client/executor/common/src/sandbox.rs

```diff
@@ -332,34 +373,115 @@ impl<FR> SandboxInstance<FR> {
 	///
 	/// The `state` parameter can be used to provide custom data for
 	/// these syscall implementations.
-	pub fn invoke<FE: SandboxCapabilities<SupervisorFuncRef=FR>>(
+	pub fn invoke<'a, FE, SCH, DTH>(
 		&self,
+
+		// function to call that is exported from the module
 		export_name: &str,
+
+		// arguments passed to the function
 		args: &[RuntimeValue],
-		supervisor_externals: &mut FE,
+
+		// arbitraty context data of the call
 		state: u32,
-	) -> std::result::Result<Option<wasmi::RuntimeValue>, wasmi::Error> {
-		with_guest_externals(
-			supervisor_externals,
-			self,
-			state,
-			|guest_externals| {
-				self.instance
-					.invoke_export(export_name, args, guest_externals)
-			},
-		)
+	) -> std::result::Result<Option<wasmi::RuntimeValue>, wasmi::Error>
+	where
+		FE: SandboxCapabilities<SupervisorFuncRef = FR> + 'a,
+		SCH: SandboxCapabiliesHolder<SupervisorFuncRef = FR, SC = FE>,
+		DTH: DispatchThunkHolder<DispatchThunk = FR>,
+	{
+		SCH::with_sandbox_capabilities( |supervisor_externals| {
+			with_guest_externals(
+				supervisor_externals,
+				self,
+				state,
+				|guest_externals| {
+					match &self.backend_instance {
+						BackendInstance::Wasmi(wasmi_instance) => {
+							let wasmi_result = wasmi_instance
+								.invoke_export(export_name, args, guest_externals)?;
+							
+							Ok(wasmi_result)
+						}
+
+						#[cfg(feature = "wasmer-sandbox")]
+						BackendInstance::Wasmer(wasmer_instance) => {
+							let function = wasmer_instance
+								.exports
+								.get_function(export_name)
+								.map_err(|error| wasmi::Error::Function(error.to_string()))?;
+
+							let args: Vec<wasmer::Val> = args
+								.iter()
+								.map(|v| match *v {
+									RuntimeValue::I32(val) => wasmer::Val::I32(val),
+									RuntimeValue::I64(val) => wasmer::Val::I64(val),
+									RuntimeValue::F32(val) => wasmer::Val::F32(val.into()),
+									RuntimeValue::F64(val) => wasmer::Val::F64(val.into()),
+								})
+								.collect();
+
+							let wasmer_result = DTH::initialize_thunk(&self.dispatch_thunk, || function.call(&args))
+								.map_err(|error| wasmi::Error::Function(error.to_string()))?;
+
+							if wasmer_result.len() > 1 {
+								return Err(wasmi::Error::Function(
+									"multiple return types are not supported yet".to_owned())
+								);
+							}
+
+							let wasmer_result = if let Some(wasmer_value) = wasmer_result.first() {
+								let wasmer_value = match *wasmer_value {
+									wasmer::Val::I32(val) => RuntimeValue::I32(val),
```

Comment:
sp_wasm_interface::Value and RuntimeValue are very alike. Is there a way to generalize type conversion?